### ISSUE TITLE
Multi-layered substrate

### ISSUE THREAD

**Issue by myurkin** [adda-team/adda#235]
After we finalize the implementation for a homogeneous semi-infinite substrate (basic surface mode) - #101, we need to think about extending it to multi-layered substrates. Here are several issues with that:
1. the easiest is to keep the limitation of particle being completely above the substrate. Then the current two-component structure of the Green's tensor can be kept.
2. the calculation of the Green's tensor will require a specialized (extended) code. However, all the optimization ideas from #176 apply.
3. The surface plasmons or guided modes may become important for the layered structures. For the standard case it is described in #234, but in the case of the dielectric waveguide, the scattered intensity at 90° will not diminish (exponentially) with r. Hence, it may require some different treatment.

The relevant papers are:
- J. Yang, J.-P. Hugonin, and P. Lalanne, “Near-to-far field transformations for radiative and guided waves,” [*ACS Photonics* **3**, 395–402 (2016)](http://doi.org/10.1021/acsphotonics.5b00559).
- A. Pors and S. I. Bozhevolnyi, “Quantum emitters near layered plasmonic nanostructures: Decay rate contributions,” [*ACS Photonics* **2**, 228–236 (2015)](http://doi.org/10.1021/ph5003472).

**Comment by myurkin** [comment]
The FFT acceleration should not require significant changes. Now we have direct interaction, which is a function of (z1-z2), and reflected part, which is a function of (z1+z2) due to the imaginary source at -z1. For multi-layered case (even two layers + vacuum) there will be infinite number of imaginary sources, but since all of them are produced by multiple reflections - they are all located at positions ai +- z1, where set of ai depends only on the geometry of the multilayered structure (widhts), but not z1. Hence, the total interaction (computed by Sommerfeld-type integrals) can be expressed as A(z1-z2) + B(z1+z2). But that means that the FFT acceleration is the same, only computation of A and B differs.

This will similarly affect optimizations for calculation of Sommerfeld integrals. Now they all depend on z1+z2, in the future we need separate path for those depending on z1-z2.

**Comment by myurkin** [comment]
Surface-plasmon polaritons (SPPs) may need separate consideration even in the case of homogeneous metallic substrate (with very low absorption). Although, they are not infinitely propagating scattering from or into SPPs may be relevant - comparable to scattering from/into slightly absorbing substrate. The relevant paper is:
Evlyukhin A.B. and Bozhevolnyi S.I. Resonant unidirectional and elastic scattering of surface plasmon polaritons by high refractive index dielectric nanoparticles, [_Phys. Rev. B_ **92**, 245419](http://doi.org/10.1103/PhysRevB.92.245419) (2015).

I am not sure what is the current limit of ADDA, when scattering angle approaches substrate from below. But most probably it is wrong (not SPP), since the attenuation constant for SPP is different from that in the bulk medium (assumed for scattering at any angle into the substrate).

**Comment by myurkin** [comment]
Here are two other projects, which struggle with the same problem with some success. So we may take some code or ideas from that:
- [SMUTHI](https://pypi.org/project/SMUTHI/)
- [IF-DDAM](https://github.com/pchaumet/IF-DDAM)

But further analysis is needed to plan further actions.

**Comment by myurkin** [comment]
Another relevant project (suggested by @AmosEgel) is https://github.com/modelics/strata . Probably we can just link to it for computation of the Green's tensor for multi-layered substrates.

**Comment by myurkin** [comment]
A development branch has been implemented some time ago by Nikita Palatkin covering two-layered substrate in the quasi-static approximation for reflected Green's tensor - https://github.com/palatni/adda/tree/multi-surface . There is no documentation except for his master thesis - see https://scattering.ru/theses/Palatkin_MSc_2022.pdf , which contains some proof-of-principle simulations, as well as unusual one like a dinosaur or a duck on top of a substrate. It is in Russian, but should be easy to automatically translate.

That implementation should be fine for sufficiently small particles and/or for large refractive index of the upper layer. Moreover, it contains implementations of the incident and scattered fields, that are general (do not require quasi-static approximation) and can be further easily generalized for substrate with any number of layers.

However, the most complicated part for a general solution to this issue remains - it is the robust and efficient calculation of Sommerfeld integrals, as thoroughly discussed above. I would argue in favor of development of specialized routine for source and observation point above the substrate (the niche, where ADDA is the most efficient). For instance, `strata` is not directly applicable to this regime, and even if some workaround is used the computed values themselves are not sufficient. We need explicit separation into two parts, depending on z1-z2 and z1+z2 respectively.

Consideration of other domains (e.g. observation point inside the particle) will be needed for, e.g., computation of near-fields or for submerged particles, but that is another story.

Finally, explicit consideration of surface plasmons are also needed, but that can be ignored, if not directly relevant for measurement at hand (either for scattering or for excitation).

**Comment by myurkin** [comment]
Efficient implementation of Sommerfeld integrals (including tabulation-interpolation) for layered medium is described in:
Pratama S. and Oosten D. van. Efficient and versatile surface integral approach to light scattering in stratified media, [_Opt. Express, OE_ **23**, 21741–21760](http://doi.org/10.1364/OE.23.021741) (2015).
